src/storage-controller/src/lib.rs

```diff
@@ -1705,76 +1709,130 @@ where
 
     async fn record_frontiers(
         &mut self,
-        external_frontiers: BTreeMap<(GlobalId, ReplicaId), Antichain<Self::Timestamp>>,
+        external_frontiers: BTreeMap<
+            GlobalId,
+            (Antichain<Self::Timestamp>, Antichain<Self::Timestamp>),
+        >,
     ) {
-        // Make `replica_id` optional, to account for storage objects that are not installed on
-        // replicas.
-        let mut frontiers = BTreeMap::new();
-        let mut external_ids = HashSet::with_capacity(frontiers.len());
-        for ((object_id, replica_id), frontier) in external_frontiers {
-            frontiers.insert((object_id, Some(replica_id)), frontier);
-            external_ids.insert(object_id);
+        let mut frontiers = external_frontiers;
+
+        // Enrich `frontiers` with storage frontiers.
+        for (object_id, collection) in self.active_collections() {
+            let since = collection.read_capabilities.frontier().to_owned();
+            let upper = collection.write_frontier.clone();
+            frontiers.insert(object_id, (since, upper));
+        }
+        for (object_id, export) in self.active_exports() {
+            // Exports cannot be read from, so their `since` is always the empty frontier.
+            let since = Antichain::new();
```

Comment:
Yeah, I'm not too happy about this either. My first version had `let since = export.read_capabitity.clone()` which I think makes even less sense. What would you prefer?

Reply:
yeah i think empty makes more sense, for now!